source/common/buffer/watermark_buffer.h

```diff
@@ -0,0 +1,64 @@
+#pragma once
+
+#include <string>
+
+#include "envoy/buffer/buffer.h"
+
+namespace Envoy {
+namespace Buffer {
+
+// A wrapper for an underlying buffer which does watermark validation.
+// The underlying buffer's ownership is transfered to the Watermark buffer.   Each time the inner
+// buffer is resized (written to or drained), the watermarks are checked.  As the buffer size
+// transitions from under the low watermark to above the high watermark, the above_high_watermark
+// function is called one time. It will not be called again until the buffer is drained below the
+// low watermark, at which point the below_low_watermark function is called.
+class WatermarkBuffer : public Instance {
```

Comment:
That's basically what I thought, but I'm happy to go either way if @mattklein123 would prefer the perf boost.
Should be up to 100% coverage on the new watermark code now.